### PR TITLE
add error mcopy test 

### PR DESCRIPTION
**Pull request by DreamWuGit**
### Description

add following tests :
1. test_oog_mcopy_max_expanded_address: expand address to max case
2. test_oog_mcopy_src_larger_dst_addr: for auditing fix mcopy expansion
3. test_oog_mcopy_max_u64_address:  u64::max edge case
4. negative test for constrain "src_address length == dst_address length" which is auditing found.

more explanation for negative test: 
currently bytecode circuit and state circuit both have negative tests(in zkevm-circuits/src/bytecode_circuit/test.rs and zkevm-circuits/src/state_circuit/test.rs). both of them take the similar way: assign overrides wrong value at assign stage.  like following code snippets 
```

fn verify_with_overrides(
    rows: Vec<Rw>,
    overrides: HashMap<(AdviceColumn, isize), Fr>,
) -> Result<(), Vec<VerifyFailure>> {
    // Sanity check that the original RwTable without overrides is valid.
    assert_eq!(verify(rows.clone()), Ok(()));

    let n_active_rows = rows.len();
    prover(rows, overrides).verify_at_rows(
        N_ROWS - n_active_rows..N_ROWS,
        N_ROWS - n_active_rows..N_ROWS,
    )
}

```
where `overrides` collection provides targeting wrong value (column + offset + wrong_value) pairs.  because of bytecode circuit and state circuit layerout are fixed, that means each column has fixed meaning.  but evm circuit layout is much more complex , besides includes multiple tables,  also include dynamic cell queries. for example, many step's witness cells are queried from some advices columns.  it is difficult to use above `overrides`  method for entire evm circuit level, plus another buss mapping module  generates most of the witness  data. 
In this PR try to  make negative test on smaller gadgets that execution step gadget take use of.  take `ErrorOOGMemoryCopyGadget` as example.
1. refactor `MemoryAddrExpandGadget` as our negative test targeting , `ErrorOOGMemoryCopyGadget` use it to construct `src_memory_addr`, `dst_memory_addr`,  `memory_expansion_mcopy`. 
2. add ErrOOGMemoryCopyGadgetTestContainer which wrapped in `UnitTestMathGadgetBaseCircuit` construction. 
3. assign wrong copy_size 
4. encounter expected constraint error : "mcopy src_address length == dst_address length" when run the circuit.

BTW, also consider another way: use feature in testing, like `#[cfg(feature = "evm_mcopy_wrong_length")]`, when such feature is enable, assign wrong value in step's `assign_exec_step` method.   it is simple and do not change circuit , but the cons is obviously not flexible, if more kinds of negative test are going to add in the future, more specific features need to be added.  

### Issue Link

[_link issue here_]

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs [hunk 1]
```diff
@@ -605,4 +706,119 @@ mod tests {
...
+            witnesses.into(),
+        );
+
+        let prover = MockProver::<Fr>::run(K as u32, &circuit, vec![]).unwrap();
+        let result = prover.verify();
+        // when not mcopy, should pass.
+        assert!(result.is_ok());
+    }
+
+    fn assert_error_matches(result: Result<(), Vec<VerifyFailure>>, name: &str) {
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs]: we can move this function to a more common file? like test_util.rs?  Can be used later.

**Review by lispc — APPROVED** [review]
clever design. Only one comment

File: zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs]: sure !

**Comment by lispc** [comment]
ToScalar triat moved to gadget crate

File: zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs]: moved [8a6e54a](https://github.com/scroll-tech/zkevm-circuits/pull/1333/commits/8a6e54a249f71240d87d95ed105eda3229545e43)
